### PR TITLE
support ignoreHidden for extract reader  && fix dead lock on dirx close

### PR DESCRIPTION
**Pull request by wonderflow**
## Changes
   
- [ ] feature1
- [ ] feature2
- [ ] fixbug1
- [ ] fixbug2
 
## Reviewers

- [ ] @[someone] please review
- [ ] @[someotherone] please review

## Wiki Changes
 
- options1...
- options2...
    
## Checklist
   
- [ ] Rebased/mergeable
- [ ] Tests pass
- [ ] Wiki updated

### CODE REVIEW

**Comment by redHJ** [comment]
ci挂了

File: reader/dirx/dir_reader.go [hunk 1]
```diff
@@ -24,6 +24,8 @@ import (
 	. "github.com/qiniu/logkit/utils/models"
 )
 
+var ErrAlreadyExist = errors.New("runner already exist")
```
Comment (redHJ) [inline reader/dirx/dir_reader.go]: 可以放在utils/models/code.go里

Comment (wonderflow) [reply reader/dirx/dir_reader.go]: 这个就当前包用到

**Comment by redHJ** [comment]
lgtm

File: reader/autofile/autofile.go [hunk 2]
```diff
@@ -51,6 +55,15 @@ func matchMode(logpath string) (path, mode string, err error) {
 		logpath = filepath.Dir(logpath)
 	}
 	path = logpath
+	if strings.HasSuffix(logpath, ".tar.gz") || strings.HasSuffix(logpath, ".tar") || strings.HasSuffix(logpath, ".zip") {
+		mode = config.ModeDirx
+		return
+	}
+	if strings.HasSuffix(logpath, ".gz") {
+		mode = config.ModeTailx
+		return
```
Comment (redHJ) [inline reader/autofile/autofile.go]: 同上

File: reader/autofile/autofile.go [hunk 3]
```diff
@@ -51,6 +55,15 @@ func matchMode(logpath string) (path, mode string, err error) {
 		logpath = filepath.Dir(logpath)
 	}
 	path = logpath
+	if strings.HasSuffix(logpath, ".tar.gz") || strings.HasSuffix(logpath, ".tar") || strings.HasSuffix(logpath, ".zip") {
+		mode = config.ModeDirx
+		return
```
Comment (redHJ) [inline reader/autofile/autofile.go]: return 具体字段

**Pull request merged**